File: src/main/kotlin/coraythan/keyswap/decks/models/DeckSearchResult.kt
package coraythan.keyswap.decks.models

import com.fasterxml.jackson.annotation.JsonInclude
import coraythan.keyswap.House
import coraythan.keyswap.auctions.DeckListing
import coraythan.keyswap.expansions.Expansion
import coraythan.keyswap.generatets.GenerateTs
import coraythan.keyswap.roundToOneSigDig
import coraythan.keyswap.synergy.SynergyCombo
import java.time.LocalDate

// It takes a long time to load all the crap in hibernate, so avoid that.
@JsonInclude(JsonInclude.Include.NON_NULL)
data class DeckSearchResult(
    val deckType: DeckType,
    val id: Long = -1,
    val keyforgeId: String = "",
    val expansion: Expansion,

    val name: String = "",

    val creatureCount: Int? = null,
    val actionCount: Int? = null,
    val artifactCount: Int? = null,
    val upgradeCount: Int? = null,

    val powerLevel: Int? = null,
    val chains: Int? = null,
    val wins: Int? = null,
    val losses: Int? = null,

    val expectedAmber: Double = 0.0,
    val amberControl: Double = 0.0,
    val creatureControl: Double = 0.0,
    val artifactControl: Double? = null,
    val efficiency: Double? = null,
    val recursion: Double? = null,
    val effectivePower: Int = 0,
    val creatureProtection: Double? = null,
    val disruption: Double? = null,
    val other: Double? = null,

    val aercScore: Int = 0,
    val previousSasRating: Int = 0,
    val previousMajorSasRating: Int? = null,
    val aercVersion: Int = 12,
    val sasRating: Int = 0,
    val synergyRating: Int = 0,
    val antisynergyRating: Int = 0,
    val metaScores: Map<String, Double> = mapOf(),
    val efficiencyBonus: Double = 0.0,

    val totalPower: Int = 0,
    val cardDrawCount: Int? = null,
    val cardArchiveCount: Int? = null,
    val keyCheatCount: Int? = null,
    val rawAmber: Int = 0,
    val totalArmor: Int? = null,

    val forSale: Boolean? = null,
    val forTrade: Boolean? = null,
    val forAuction: Boolean? = null,
    val wishlistCount: Int? = null,
    val funnyCount: Int? = null,

    val lastSasUpdate: String? = null,
    val sasPercentile: Double = 0.0,

    val hasOwnershipVerification: Boolean? = null,

    val housesAndCards: List<HouseAndCards> = listOf(),
    val deckSaleInfo: List<DeckSaleInfo>? = null,
    val owners: List<String>? = null,
    val synergyDetails: List<SynergyCombo>? = null,
    val dateAdded: LocalDate? = null,
    val twinId: String? = null,
    val tokenInfo: TokenInfo? = null,

    val allianceHouses: List<AllianceHouseInfo>? = null,
    val validAlliance: Boolean? = null,
    val discoveredBy: String? = null,
) {

    fun printDeckSimple(saleInfo: DeckListing? = null): String {

        val buyItNow = saleInfo?.buyItNow
        val highBid = saleInfo?.highestBid
        val highOffer = saleInfo?.highestOffer
        val currencySymbol = saleInfo?.currencySymbol
        val buyItNowMessage = if (buyItNow == null) "" else ", BIN: $currencySymbol$buyItNow"

        val forSaleMessage = when {
            forAuction == true -> " • On Auction${if (highBid == null) "" else ", high bid: $currencySymbol$highBid"}$buyItNowMessage"
            forSale == true -> " • For Sale${if (highOffer == null) "" else ", high offer:  $currencySymbol$highOffer"}$buyItNowMessage"
            else -> ""
        }

        val tokenName = if (this.tokenInfo == null) "" else " • ${this.tokenInfo.name}"

        return "$sasRating SAS • ${expansion.readable} • ${
            housesAndCards.map { it.house }.joinToString(" – ") { it.masterVaultValue }
        }" + tokenName + forSaleMessage
    }

    fun printDeck(saleInfo: DeckListing? = null): String {
        return """
            ${printDeckSimple(saleInfo)}
            ${amberControl.roundToOneSigDig()} A • ${expectedAmber.roundToOneSigDig()} E • ${(artifactControl ?: 0.0).roundToOneSigDig()} R • ${creatureControl.roundToOneSigDig()} C
            ${(efficiency ?: 0.0).roundToOneSigDig()} F • ${(recursion ?: 0.0).roundToOneSigDig()} U • ${(disruption ?: 0.0).roundToOneSigDig()} D
        """.trimIndent()
    }

    fun toSimpleResult() = SimpleDeckSearchResult(
        name,
        keyforgeId,
        housesAndCards.map { it.house },
        sasRating
    )
}

@GenerateTs
data class SimpleDeckSearchResult(
    val name: String,
    val keyforgeId: String,
    val houses: List<House> = listOf(),
    val sas: Int = 0,
)

@GenerateTs
data class AllianceHouseInfo(
    val keyforgeId: String,
    val name: String,
    val house: House,
)

@GenerateTs
data class TokenInfo(
    val id: String,
    val name: String,
    val house: House,
)

File: ui/src/decks/models/DeckSearchResult.ts
import {DeckSaleInfo} from "../../generated-src/DeckSaleInfo"
import {Expansion} from "../../generated-src/Expansion"
import {House} from "../../generated-src/House"
import {HouseAndCards} from "../../generated-src/HouseAndCards"
import {SynergyCombo} from "../../generated-src/SynergyCombo"
import {SynergyTrait} from "../../generated-src/SynergyTrait"
import {DeckType} from "../../generated-src/DeckType"
import {AllianceHouseInfo} from "../../generated-src/AllianceHouseInfo"
import {TokenInfo} from "../../generated-src/TokenInfo"
import {TokenCreationValues} from "../../generated-src/TokenCreationValues"

export interface DeckWithSynergyInfo {
    deck: DeckSearchResult
    cardRatingPercentile: number
    synergyPercentile: number
    antisynergyPercentile: number
    alliance: boolean
}

export interface DeckSearchResult {

    deckType: DeckType
    id: number
    keyforgeId: string
    expansion: Expansion

    name: string

    creatureCount?: number
    actionCount?: number
    artifactCount?: number
    upgradeCount?: number

    powerLevel?: number
    chains?: number
    wins?: number
    losses?: number

    aercScore: number
    previousSasRating?: number
    previousMajorSasRating?: number
    sasRating: number
    synergyRating: number
    antisynergyRating: number
    efficiencyBonus: number

    totalPower: number
    cardDrawCount?: number
    cardArchiveCount?: number
    keyCheatCount?: number
    rawAmber: number
    totalArmor?: number

    forSale?: boolean
    forTrade?: boolean
    wishlistCount?: number
    funnyCount?: number

    lastSasUpdate?: string
    sasPercentile: number

    housesAndCards: HouseAndCards[]
    deckSaleInfo?: DeckSaleInfo[]
    owners?: string[]
    synergyDetails?: SynergyCombo[]
    tokenCreationValues?: TokenCreationValues
    hasOwnershipVerification: boolean
    dateAdded?: string

    amberControl: number
    expectedAmber: number
    artifactControl?: number
    creatureControl: number
    efficiency?: number
    recursion?: number
    effectivePower: number
    creatureProtection?: number
    disruption?: number
    other?: number

    twinId?: string
    tokenInfo?: TokenInfo
    hauntingOdds?: number

    allianceHouses?: AllianceHouseInfo[]
    validAlliance?: boolean
    discoveredBy?: string
}

export interface DeckIdentifyingInfo {
    id: number
    keyforgeId: string
    deckType: DeckType
    name: string
}

export interface TraitInDeckInfo {
    trait: SynergyTrait
    count: number
    strength: number
    houseCounts: Map<House, number>
}

export interface DeckPage {
    decks: DeckSearchResult[]
    page: number
}

export interface DeckCount {
    pages: number
    count: number
}
